File: Python/ByRound/2127/2127_D_Root_was_Built_by_Love_Broken_by_Destiny.py
# Problem: CF 2127 D - Root was Built by Love, Broken by Destiny
# https://codeforces.com/contest/2127/problem/D

"""
D. Root was Built by Love, Broken by Destiny

Purpose:
This code solves a graph-based combinatorial problem where we need to count the number of valid ways to assign
houses to two sides of a river such that all bridges connect houses on opposite sides and do not cross when
drawn as straight lines.

Key algorithms and techniques used:
- Graph theory (connected components, degree analysis)
- Tree structure recognition (m == n-1 implies tree)
- Factorial computation with modular arithmetic
- Combinatorial counting using degrees and leaf nodes

Time Complexity: O(n + m) per test case, where n is number of houses and m is number of bridges.
Space Complexity: O(n + m) for adjacency list and degree arrays.

"""

import sys

MOD = 10**9 + 7
MAXN = 2 * 10**5 + 5
fact = [1] * MAXN
for i in range(1, MAXN):
    fact[i] = (fact[i - 1] * i) % MOD


def solve():
    it = iter(map(int, sys.stdin.buffer.read().split()))
    t = next(it)
    out_lines = []
    for _ in range(t):
        n = next(it)
        m = next(it)
        adj = [[] for _ in range(n)]
        deg = [0] * n
        for _ in range(m):
            u = next(it) - 1
            v = next(it) - 1
            adj[u].append(v)
            adj[v].append(u)
            deg[u] += 1
            deg[v] += 1
        # Check if graph is a tree (necessary condition for valid arrangement)
        if m != n - 1:
            out_lines.append("0")
            continue
        ans = 2  # Two possible sides for first node
        leaves = sum(1 for d in deg if d == 1)  # Count leaf nodes (degree 1)
        ok = True
        # Process each non-leaf node
        for u in range(n):
            if deg[u] > 1:
                x = 0
                # Count neighbors of degree 1 (leaves)
                for v in adj[u]:
                    if deg[v] == 1:
                        x += 1
                # If number of leaf neighbors is less than (degree - 2), invalid arrangement
                if x < deg[u] - 2:
                    ok = False
                    break
                # Multiply answer by factorial of leaf neighbors (permutations)
                ans = (ans * fact[x]) % MOD
        if not ok:
            out_lines.append("0")
            continue
        # If there are fewer leaves than n-1, then we can permute more
        if leaves < n - 1:
            ans = (ans * 2) % MOD
        out_lines.append(str(ans))
    sys.stdout.write("\n".join(out_lines))


if __name__ == "__main__":
    solve()


# https://github.com/VaHiX/codeForces/

File: Java/ByRound/2127/2127_D_Root_was_Built_by_Love_Broken_by_Destiny.java
// Problem: CF 2127 D - Root was Built by Love, Broken by Destiny
// https://codeforces.com/contest/2127/problem/D

/*
 * Problem: D. Root was Built by Love, Broken by Destiny
 *
 * Purpose: 
 *   Given a graph with n nodes (houses) and m edges (bridges), calculate the number
 *   of valid ways to arrange houses on two sides of a river such that:
 *   1. Every bridge connects houses on opposite sides.
 *   2. Bridges do not cross when drawn as straight lines.
 *
 * Approach:
 *   - The graph must form a tree (m = n - 1) for it to be valid, otherwise return 0.
 *   - Find the "core" nodes (nodes with degree >= 2), which form a path in the tree.
 *   - If there are 0 core nodes: answer is 2 (assign sides to each house).
 *   - If there is 1 core node: answer is 2 * (n-1)!.
 *   - Otherwise:
 *     - The core graph must form a path (degree of each core node is at most 2).
 *     - Count how many leaves are attached to each core node (in the core subgraph).
 *     - Total arrangements = (product of factorials of leaf counts) * 4.
 *
 * Time Complexity: O(n + m)
 * Space Complexity: O(n + m)
 */

import java.io.*;
import java.util.*;

public class Main {
    static final int MOD = (int)1e9 + 7;
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int t = Integer.parseInt(st.nextToken());
        int[] fact = new int[200000 + 5];
        factorial(fact, MOD);
        while (t-- > 0) {
            st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            int m = Integer.parseInt(st.nextToken());
            
            // Build adjacency list representation of the graph
            @SuppressWarnings("unchecked")
            ArrayList<Integer>[] graph = (ArrayList<Integer>[]) new ArrayList[n];
            for (int i = 0; i < n; i++) graph[i] = new ArrayList<>();
            for (int i = 0; i < m; i++) {
                st = new StringTokenizer(br.readLine());
                int u = Integer.parseInt(st.nextToken()) - 1;
                int v = Integer.parseInt(st.nextToken()) - 1;
                graph[u].add(v);
                graph[v].add(u);
            }
            
            // Check if graph is a tree
            if (m != n - 1 || !isConnected(graph, n)) {
                System.out.println(0);
                continue;
            }
            
            // Compute degree of each node
            int[] deg = new int[n];
            for (int i = 0; i < n; i++) deg[i] = graph[i].size();
            
            // Identify core nodes (nodes with degree >= 2)
            boolean[] isCore = new boolean[n];
            int coreCount = 0;
            for (int i = 0; i < n; i++) {
                if (deg[i] >= 2) {
                    isCore[i] = true;
                    coreCount++;
                }
            }
            
            // If no core nodes, we can assign sides arbitrarily to each pair of houses
            if (coreCount == 0) {
                System.out.println(2);
                continue;
            }
            
            // If exactly one core node, it's just a star graph -> 2 * (n-1)!
            if (coreCount == 1) {
                long ans = (long)2 * fact[n - 1] % MOD;
                System.out.println(ans);
                continue;
            }

            // Count degree of each node in core subgraph
            int[] coreDeg = new int[n]; 
            for (int u = 0; u < n; u++) if (isCore[u]) {
                for (int v : graph[u]) if (isCore[v]) {
                    coreDeg[u]++;
                }
            }

            boolean ok = true;
            int endpoints = 0;
            int startCore = -1;
            for (int i = 0; i < n; i++) if (isCore[i]) {
                // If a core node has degree > 2, invalid arrangement
                if (coreDeg[i] > 2) { ok = false; break; }
                // Count the number of endpoints (degree 1)
                if (coreDeg[i] == 1) { endpoints++; startCore = i; }
                // If no start yet but found a degree 2 node, use it
                if (startCore == -1 && coreDeg[i] == 2) startCore = i;
            }

            // Invalid if any core node has degree > 2 or not exactly two endpoints
            if (!ok || endpoints != 2) {
                System.out.println(0);
                continue;
            }

            // BFS to check connectivity among core nodes
            int seenCore = bfsCoreReachableCount(graph, isCore, startCore);
            if (seenCore != coreCount) {
                System.out.println(0);
                continue;
            }
            
            // For each core node u, count how many leaves are attached to it
            long ways = 1;
            for (int u = 0; u < n; u++) if (isCore[u]) {
                int leavesAtU = deg[u] - coreDeg[u]; 
                ways = (ways * fact[leavesAtU]) % MOD;
            }
            
            // Multiply by 4: one of the endpoints can be flipped
            ways = (ways * 4) % MOD;
            System.out.println(ways);
        }
    }

    // Check if graph is connected using BFS
    private static boolean isConnected(ArrayList<Integer>[] g, int n) {
        boolean[] vis = new boolean[n];
        ArrayDeque<Integer> dq = new ArrayDeque<>();
        dq.add(0);
        vis[0] = true;
        int seen = 1;
        while (!dq.isEmpty()) {
            int u = dq.poll();
            for (int v : g[u]) if (!vis[v]) {
                vis[v] = true;
                seen++;
                dq.add(v);
            }
        }
        return seen == n;
    }

    // BFS to count number of reachable core nodes from a starting core node
    private static int bfsCoreReachableCount(ArrayList<Integer>[] g, boolean[] isCore, int start) {
        boolean[] vis = new boolean[g.length];
        ArrayDeque<Integer> dq = new ArrayDeque<>();
        dq.add(start);
        vis[start] = true;
        int cnt = 1;
        while (!dq.isEmpty()) {
            int u = dq.poll();
            for (int v : g[u]) if (isCore[v] && !vis[v]) {
                vis[v] = true;
                cnt++;
                dq.add(v);
            }
        }
        return cnt;
    }

    // Precompute factorials modulo MOD
    private static void factorial(int[] fact, int mod) {
        fact[0] = 1;
        for (int i = 1; i < fact.length; i++) {
            fact[i] = (int)(((long)fact[i - 1] * i) % mod);
        }
    }
}


// https://github.com/VaHiX/CodeForces/